### PR TITLE
Allow link_depends to take strings, Files or generated objects. Close…

### PR DESCRIPTION
**Pull request by dcbaker**
…s #1172

Currently only strings can be passed to the link_depends argument of
executable and *library, which solves many cases, but not every one.
This patch allows generated sources and Files to be passed as well.

On the implementation side, it uses a helper method to keep the more
complex logic separated from the __init__ method.

### CODE REVIEW

**Comment by aaronp24** [comment]
This doesn't seem to work for custom targets. I'm using this as my test: https://github.com/aaronp24/libglvnd/commit/809c011014aec22c406904e8dcebc6b0e35d8636

The rules in question are these:
```
glapi_gen_libopengl_exports_script = find_program('gen_libOpenGL_exports.py')
glapi_gen_gl_core_xml = files('xml/gl.xml')
g_opengl_exports_sym = custom_target('g_opengl_exports.sym',
    command : [glapi_gen_libopengl_exports_script, 'opengl', glapi_gen_gl_core_xml],
    capture : true,
    depend_files : files('genCommon.py', 'gen_libOpenGL_exports.py') + glapi_gen_gl_core_xml,
    output : 'g_opengl_exports.sym',
)
g_opengl_exports_sym_dep = declare_dependency(
    sources : g_opengl_exports_sym,
    link_args : '-Wl,--version-script,@0@'.format(g_opengl_exports_sym.full_path()),
)

opengl = shared_library('OpenGL', 'libopengl.c',
    link_depends : [
        g_opengl_exports_sym,
    ],
    link_whole : [
        glapi_opengl,
    ],
    dependencies : [
        gldispatch_dep,
        g_opengl_exports_sym_dep,
    ],
    include_directories : [inc, util_inc],
    version : '0.0.0',
    install : true,
)
```

I added some tracing to process_link_depends to print out the entries here, and this thing is a CustomTargetHolder with a CustomTarget that apparently has a get_generated_sources, but where s.get_generated_sources() returns []. So it doesn't complain, but it also doesn't create a dependency on the linker script.

**Comment by dcbaker** [comment]
You're right. `get_outputs()` returns what we actually want. Of course meson is so short on comments it's hard to tell what each one does and why to use it :/

**Comment by aaronp24** [comment]
Still doesn't work. This sticks g_opengl_exports.sym into the dependency list, which somehow turns into /home/aaron/git/libglvnd/src/OpenGL/g_opengl_exports.sym in build.ninja. But the generated file is /tmp/b/src/generate/g_opengl_exports.sym

**Comment by dcbaker** [comment]
yup. I just noticed that. unclean build working, clean build not so much.

I think what I really want to do is add a link_script keyword argument, like the `vs_module_defs` parameter, which could handle both generating the -Wl,--version-script and the dependency anyway.

**Comment by dcbaker** [comment]
@aaronp24 I think this version should work for you. I'll see about adding an extra kwarg to make this even easier.

**Comment by aaronp24** [comment]
Nope. That creates a dependency on src/OpenGL/g_opengl_exports.sym, but the generated file is in src/generate/g_opengl_exports.sym:

```
$ ninja -C /tmp/b src/generate/g_opengl_exports.sym
ninja: Entering directory `/tmp/b'
[1/1] 'Generating g_opengl_exports.sym with a meson_exe.py custom command.'
$ ninja -C /tmp/b
ninja: Entering directory `/tmp/b'
ninja: error: 'src/OpenGL/g_opengl_exports.sym', needed by 'src/OpenGL/libOpenGL.so.0.0.0', missing and no known rule to make it
```

I could probably work around that somehow, by moving files around, I guess, but it seems confusing.

**Comment by dcbaker** [comment]
No, this should happen transparently. I'll poke at it some more after lunch.

**Comment by dcbaker** [comment]
Okay, it was a really trivial bug on my part. I've tested glvnd with this version of the patch, and it works (I do have a very small glvnd patch.

**Comment by aaronp24** [comment]
Thanks, that seems to work!

The libglvnd thing is just an experiment in progress, so I'm definitely open to suggestions.

**Comment by dcbaker** [comment]
I figured. I'm trying to port mesa ATM, so having GLVND ported to meson too would be pretty nifty.

**Comment by jpakkane** [comment]
This message is posted to every outstanding merge request.

We have transitioned from Github wiki to our new documentation system that generates all documentation directly from the Git repo. This means that from now on every merge request must come with corresponding documentation changes.

If your MR requires documentation changes, do the necessary changes to files in `docs/markdown`.

If your MR adds a new feature, add a description to the release notes in `docs/markdown/Release-notes-for-0.41.0.md`.

If this commit is just a bugfix with no functional changes, you don't need to do anything.

Thank you

**Comment by jpakkane** [comment]
Traditionally Meson's implementation has been a mix of File objects and strings. We are trying to move closer to the case where all entries that are files (whether source or generated) are only stored in File objects. This means converting strings to Files as soon as possible and only going back to strings when serializing in the backend. See for example #1706. Having dependencies as strings rather than objects makes many things needlessly complicated (a fair amount of the terribleness of `ninjabackend.py` is caused by this).

`process_link_depends` currently seems to convert file objects to strings when it should be doing the opposite.

File: mesonbuild/build.py [hunk 1]
```diff
@@ -505,6 +505,35 @@ def validate_sources(self):
...
+        for s in sources:
+            if hasattr(s, 'held_object'):
+                s = s.held_object
+
+            if isinstance(s, File):
+                self.link_depends.append(s)
+            elif isinstance(s, str):
+                self.link_depends.append(
+                    File.from_source_file(environment.source_dir, self.subdir, s))
+                    # os.path.join(environment.source_dir, self.subdir, s))
```
Comment (ignatenkobrain) [inline mesonbuild/build.py]: __[flake8]__
* [E116] unexpected indentation (comment)

File: mesonbuild/build.py [hunk 2]
```diff
@@ -505,6 +505,35 @@ def validate_sources(self):
...
+            elif hasattr(s, 'get_outputs'):
+                self.link_depends.extend(
+                    [File.from_built_file(s.subdir, p) for p in s.get_outputs()])
+                    # [os.path.join(s.subdir, p) for p in s.get_outputs()])
```
Comment (ignatenkobrain) [inline mesonbuild/build.py]: __[flake8]__
* [E116] unexpected indentation (comment)

**Comment by dcbaker** [comment]
Updated `link_depends` to always be a `File` instead of a string. It does require some changes to the ninja backend (and possible the VS backends, but I haven't seen those come back yet)

**Comment by dcbaker** [comment]
No VS backend changes appear to be necessary. Anything else I should change, or is this ready to merge?

File: mesonbuild/backend/ninjabackend.py [hunk 3]
```diff
@@ -2336,6 +2336,12 @@ def determine_rpath_dirs(self, target):
     def get_dependency_filename(self, t):
         if isinstance(t, build.SharedLibrary):
             return os.path.join(self.get_target_private_dir(t), self.get_target_filename(t) + '.symbols')
+        elif isinstance(t, mesonlib.File):
+            if t.is_built:
+                return t.relative_name()
+            else:
+                return t.absolute_path(self.environment.get_source_dir(),
+                                       self.environment.get_build_dir())
```
Comment (jpakkane) [inline mesonbuild/backend/ninjabackend.py]: Add an else that throws unreachable code exception here.

File: mesonbuild/build.py [hunk 4]
```diff
@@ -505,6 +505,33 @@ def validate_sources(self):
                 assert(len(self.compilers) == 1)
                 return
 
+    def process_link_depends(self, sources, environment):
+        """Process the link_depends keyword argument.
+
+        This is designed to handle strings, Files, and the output of Custom
+        Targets. Notably it doesn't handle generator() returned objects. Since
+        adding them as a link depends would inherently cuase them to be
```
Comment (jpakkane) [inline mesonbuild/build.py]: cuase -> cause

File: mesonbuild/backend/ninjabackend.py (hunk 3 above)
Comment (dcbaker) [inline mesonbuild/backend/ninjabackend.py]: to the `if t.isbuilt` if, or the other one?

Comment (jpakkane) [inline mesonbuild/backend/ninjabackend.py]: The outer one. If there is an item that is neither of the tested things, it is silently discarded. That is not good.

Comment (dcbaker) [inline mesonbuild/backend/ninjabackend.py]: It gets passed to `self.get_target_filename()` if it isn't handled in that if block, which is what it did before my patch. Is that not correct?

Comment (jpakkane) [inline mesonbuild/backend/ninjabackend.py]: My bad, got cross-eyed.

Comment (dcbaker) [inline mesonbuild/backend/ninjabackend.py]: No worries, happens to me all the time.

**Comment by dcbaker** [comment]
Is there anything missing for this to land?

**Comment by jon-turney** [comment]
> No VS backend changes appear to be necessary.

This seems to be because VS backend has always been ignoring link_depends.  Of course, now that may contain generated files, this is a problem, see #1799

**Pull request merged**